Source/DafnyLanguageServer/MarkupTestFile.cs

```diff
@@ -0,0 +1,237 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable disable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Dafny.LanguageServer.Workspace;
+using OmniSharp.Extensions.LanguageServer.Protocol.Models;
+using Range = OmniSharp.Extensions.LanguageServer.Protocol.Models.Range;
+
+namespace Microsoft.Dafny.LanguageServer {
+  /// <summary>
+  /// To aid with testing, we define a special type of text file that can encode additional
+  /// information in it.  This prevents a test writer from having to carry around multiple sources
+  /// of information that must be reconstituted.  For example, instead of having to keep around the
+  /// contents of a file *and* and the location of the cursor, the tester can just provide a
+  /// string with the "$" character in it.  This allows for easy creation of "FIT" tests where all
+  /// that needs to be provided are strings that encode every bit of state necessary in the string
+  /// itself.
+  /// 
+  /// The current set of encoded features we support are: 
+  /// 
+  /// $$ - The position in the file.  There can be at most one of these.
+  /// 
+  /// [| ... |] - A span of text in the file.  There can be many of these and they can be nested
+  /// and/or overlap the $ position.
+  /// 
+  /// {|Name: ... |} A span of text in the file annotated with an identifier.  There can be many of
+  /// these, including ones with the same name.
+  /// 
+  /// Additional encoded features can be added on a case by case basis.
+  /// </summary>
+  public static class MarkupTestFile {
+    private const string PositionString = "$$";
+    private const string SpanStartString = "[|";
+    private const string SpanEndString = "|]";
+    private const string NamedSpanStartString = "{|";
+    private const string NamedSpanEndString = "|}";
+
+    private static readonly Regex s_namedSpanStartRegex = new Regex(@"\{\| ([-_.A-Za-z0-9\+]+) \:",
+        RegexOptions.Multiline | RegexOptions.IgnorePatternWhitespace);
+
+    private static void Parse(
+        string input, out string output, out List<int> positions, out IDictionary<string, List<TextSpan>> spans) {
+      positions = new List<int>();
+      var tempSpans = new Dictionary<string, List<TextSpan>>();
+
+      var outputBuilder = new StringBuilder();
+
+      var currentIndexInInput = 0;
+      var inputOutputOffset = 0;
+
+      // A stack of span starts along with their associated annotation name.  [||] spans simply
+      // have empty string for their annotation name.
+      var spanStartStack = new Stack<(int matchIndex, string name)>();
+      var namedSpanStartStack = new Stack<(int matchIndex, string name)>();
+
+      while (true) {
```

Comment:
So the current algorithm is the following:
* Start with index 0
* Collect the first index at which there is a match for each of $$, [|, |], |}, or {|name:
* Find the first match. If there is an ambiguity such as [|], then try to close existing opened spans or named spans first..
* If it's a position, we add it to the position
* If we can close a span or named span, then we add it. If it's an open span or named span, we put it on a stack.
* In any case, we keep an offset to always keep the right position.

I see that you need to do something like this because you can have positions inside spans, so that makes sense.

However, could you envision the following cleaner code:
* Use `{{name:`, `[[`, `]]`, `}}` or another markup chosen by you so that you just cannot have the `[|]` or `{|]` ambiguities. I also though of `[<` and `>]` and have the positions marked like a cross like `><`, but feel free to use your own ideas.
* Make use of a regular expression that avoids running match four times and one extra time for the regular expression. I'm not concerned about performance, since these are small tests, but more about maintenance::

```
var r = new Regex(@"(?<Position>\$\$)|(?<SpanStart>\[\|)|(?<SpanEnd>\|\])|(?<NameSpanStart>\{\| ([-_.A-Za-z0-9\+]+) \:)|(?<NameSpanEnd>\|\})");

var lastIndex = 0;
foreach(var match in r.matches(input)) {
  if(match.Groups["Position"].Success) {
    ....
  } else if(match.Groups["SpanStart"].Success) {
  } ....

  // Store the string between lastIndex and match.Index
  // Reassign lastIndex
  var lastIndex = match.Index + match.Groups[0].Value.Length;
  // And other offset computations to keep track of positions
}
```

Reply:
> I also though of [< and >] and have the positions marked like a cross like ><, but feel free to use your own ideas.

I like that. Is the ambiguity a problem that needs to be solved though? I'm hesitant about changing the code since I took this from Roslyn so I'm confident it works and won't need much maintenance.

Reply:
The fact that 15 lines of code are dedicated to solve an ambiguity which has no reason to exist other than an arbitrary design is kind-of bothering me :-) But do you think the proposed fix for the ambiguity requires too much work?

I'm also surprised that Roslyn provided such code that seems underperforming at best (O(N*M) where N is the length of the code and M is the number of spans/positions), but moreover it was very confusing to me when I reviewed it - I kept asking myself why do we do all these matches as well as one regex? Why sorting a list? Why implementing in a worse complexity what a simpler regular expression can do?

Do you think the version I suggested with regex.matches might be less clear, too long or less maintainable ?
I know you already did a lot of effort there so I don't want to ask you too much.

For me, implementing it with regex.matches will make me much more comfortable to approve it than the current code.

Reply:
Made the change. I picked `[>...<]` and `{>...<}` over `[<...>]` and `{<...>}` because I found it more visually appealing.